File: 0100-same-tree/0100-same-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool isSameTree(TreeNode* p, TreeNode* q) {
        // // if both are null then they are identical, there is no point of difference
        // if(p == NULL && q == NULL)return true;
        // // if any one of them is null then false
        // if(p == NULL || q == NULL)return false;

        // // both are not null
        // if((p->val) != (q->val))return false;

        // return (isSameTree(p->left,q->left)&&isSameTree(p->right,q->right));



        if(p == NULL || q == NULL)return (p == q);

        return ((p->val == q->val)&&isSameTree(p->left,q->left)&&isSameTree(p->right,q->right));
    }
};

File: 0100-same-tree/0100-same-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
 // O(n)
class Solution {
    public boolean isSameTree(TreeNode p, TreeNode q) {
        if(p == null || q == null)return (p == q);// this one is enough
        // if(p == null && q != null)return false;
        // if(p != null && q == null)return false;
        // if(p == null && q == null)return true;
        // if(p.val != q.val)return false;

        return (p.val == q.val && isSameTree(p.left,q.left) && isSameTree(p.right,q.right));
    }
}